src/plugins.ts

```diff
@@ -0,0 +1,190 @@
+	create(createInfo: PluginCreateInfo): ts.LanguageService;
+	getExternalFiles?(proj: Project): string[];
+}
+
+/**
+ * All of tsserver's environment exposed to plugins
+ */
+export interface PluginCreateInfo {
+	project: Project;
+	languageService: ts.LanguageService;
+	languageServiceHost: ts.LanguageServiceHost;
+	serverHost: ServerHost;
+	config: any;
+}
+
+/**
+ * The portion of tsserver's Project API exposed to plugins
+ */
+export interface Project {
+	projectService: {
+		logger: Logger;
+	};
+}
+
+/**
+ * The portion of tsserver's ServerHost API exposed to plugins
+ */
+export type ServerHost = object;
+
+/**
+ * The result of a node require: a module or an error.
+ */
+type RequireResult = { module: {}, error: undefined } | { module: undefined, error: {} };
+
+export class PluginLoader {
+
+	private allowLocalPluginLoads: boolean = false;
+	private globalPlugins: string[] = [];
+	private pluginProbeLocations: string[] = [];
+
+	constructor(
+		private rootFilePath: string,
+		private fs: ts.ModuleResolutionHost,
+		pluginSettings?: PluginSettings,
+		private logger = new NoopLogger(),
+		private resolutionHost = new LocalModuleResolutionHost(),
+		private requireModule: (moduleName: string) => any = require) {
+		if (pluginSettings) {
+			this.allowLocalPluginLoads = pluginSettings.allowLocalPluginLoads || false;
+			this.globalPlugins = pluginSettings.globalPlugins || [];
+			this.pluginProbeLocations = pluginSettings.pluginProbeLocations || [];
+		}
+	}
+
+	public loadPlugins(options: ts.CompilerOptions, applyProxy: EnableProxyFunc) {
+		// Search our peer node_modules, then any globally-specified probe paths
+		// ../../.. to walk from X/node_modules/javascript-typescript-langserver/lib/project-manager.js to X/node_modules/
+		const searchPaths = [combinePaths(__filename, '../../..'), ...this.pluginProbeLocations];
+
+		// Corresponds to --allowLocalPluginLoads, opt-in to avoid remote code execution.
+		if (this.allowLocalPluginLoads) {
+			const local = this.rootFilePath;
+			this.logger.info(`Local plugin loading enabled; adding ${local} to search paths`);
+			searchPaths.unshift(local);
+		}
+
+		let pluginImports: ts.PluginImport[] = [];
+		if (options.plugins) {
+			pluginImports = options.plugins as ts.PluginImport[];
+		}
+
+		// Enable tsconfig-specified plugins
+		if (options.plugins) {
+			for (const pluginConfigEntry of pluginImports) {
+				this.enablePlugin(pluginConfigEntry, searchPaths, applyProxy);
+			}
+		}
+
+		if (this.globalPlugins) {
+			// Enable global plugins with synthetic configuration entries
+			for (const globalPluginName of this.globalPlugins) {
+				// Skip already-locally-loaded plugins
+				if (!pluginImports || pluginImports.some(p => p.name === globalPluginName)) {
+					continue;
+				}
+
+				// Provide global: true so plugins can detect why they can't find their config
+				this.enablePlugin({ name: globalPluginName, global: true } as ts.PluginImport, searchPaths, applyProxy);
```

Comment:
why is `global` not in the typings?

Reply:
Code is copied from TS project - fix should be made there